rust-runtime/aws-smithy-http-server/src/plugin/closure.rs

```diff
@@ -0,0 +1,29 @@
+use tower::layer::util::Stack;
+
+use crate::operation::{Operation, OperationShape};
+
+use super::Plugin;
+
+/// A [`Plugin`] implemented by a closure over the operation name. See [`plugin_from_operation_name_fn`] for more details.
+pub struct OperationNameFn<F> {
+    f: F,
+}
+
+impl<P, Op, S, ExistingLayer, NewLayer, F> Plugin<P, Op, S, ExistingLayer> for OperationNameFn<F>
+where
+    F: Fn(&'static str) -> NewLayer,
+    Op: OperationShape,
+{
+    type Service = S;
+    type Layer = Stack<ExistingLayer, NewLayer>;
+
+    fn map(&self, input: Operation<S, ExistingLayer>) -> Operation<Self::Service, Self::Layer> {
+        input.layer((self.f)(Op::NAME))
+    }
+}
+
+/// Constructs a [`Plugin`] using a closure over the operation name `F: Fn(&'static str) -> L` where `L` is a HTTP
+/// [`Layer`](tower::Layer).
+pub fn plugin_from_operation_name_fn<F>(f: F) -> OperationNameFn<F> {
```

Comment:
Would it be useful to add the constraint `F: Fn(&'static str) -> NewLayer` here to force earlier compile errors on wrong usage?

Reply:
The reason I haven't added it here is because [layer_fn](https://docs.rs/tower/latest/tower/layer/fn.layer_fn.html) doesn't have these bounds and, prior to [`fn-traits`](https://doc.rust-lang.org/beta/unstable-book/library-features/fn-traits.html) I think we'd have to add a second generic on here `<L, F>` to specify the return type of `Fn(&'static str) -> L`.

Reply:
Yeah, we would have to add an `L` parameter - but I think it's worth it for the more precise error message.